### PR TITLE
Switch children aggregation to registerAggregation and remove its PROTOTYPE

### PR DESCRIPTION
**Pull request by nik9000**
This required some invasive surgery on the serialization but it ought to give us a way to do one aggregation at a time.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/search/aggregations/BaseAggregationTestCase.java [hunk 1]
```diff
@@ -253,13 +253,12 @@ public void testFromXContent() throws IOException {
     public void testSerialization() throws IOException {
         AB testAgg = createTestAggregatorBuilder();
         try (BytesStreamOutput output = new BytesStreamOutput()) {
-            testAgg.writeTo(output);
+            output.writeAggregatorBuilder(testAgg);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/aggregations/BaseAggregationTestCase.java]: I wanted to test that names came out right.

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorBuilder.java [hunk 2]
```diff
@@ -82,6 +82,93 @@ protected ValuesSourceAggregatorBuilder(String name, Type type, ValuesSourceType
     }
 
     /**
+     * Read from a stream.
+     */
+    protected ValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType)
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorBuilder.java]: I know sometimes `valuesSourceType` and `targetValueType` have to be serialized. I think we can handle that in the PR for that builder. I'm confident we can serialize it if we need it though.

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorBuilder.java [hunk 3]
```diff
@@ -62,6 +62,47 @@ public AggregatorBuilder(String name, Type type) {
...
+     * Read from a stream.
+     */
+    protected AggregatorBuilder(StreamInput in, Type type) throws IOException {
+        name = in.readString();
+        this.type = type;
+        factoriesBuilder = AggregatorFactories.Builder.PROTOTYPE.readFrom(in);
+        metaData = in.readMap();
+    }
+
+    protected boolean usesNewStyleSerialization() {
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorBuilder.java]: Presumably this should be removed when we have migrated all the builders to the new serialisation? If so, can we add a NORELEASE comment so we don't forget?

**Comment by colings86** [comment]
LGTM. I left one minor comment. The `usesNewStyleSerialization` stuff isn't so nice but I know why you had to do it that way and I'm fine with it as a temporary thing. It will be good when all the aggregations are moved over and we can remove this.

**Comment by nik9000** [comment]
Thanks for the review @colings86 ! I'm going to merge this and try and do one of the more complicated ones next.

**Comment by colings86** [comment]
Cool, I would suggest doing something like the Range or Histogram Aggregation next if you want one of the more complex ones (although hopefully it will still be quite simple to do)

**Comment by nik9000** [comment]
> Range or Histogram Aggregation

Cool! I'm just waiting on the tests to pass one last time. This is one of those changes where one misplaced `}` could cause sneaky problems. The `usesNewStyleSerialization` hack would be unpleasant to keep for very long....

**Pull request merged**